src/sentry/utils/services.py

```diff
@@ -76,3 +83,173 @@ def expose(self, context):
+        },
+        # ... etc ...
+
+    The backends used for a method call are determined by a selector function.
+    The return value of the selector function is a list of strings, which
+    correspond to names in the backend mapping. The first item in the result
+    list is considered the "primary backend". The remainder of the items in the
+    result list are considered "secondary backends". The result value of the
+    primary backend will be the result value of the delegated method (to
+    callers, this appears as a synchronous method call.) The secondary backends
+    are called asynchronously in the background. (To recieve the result values
+    of these method calls, provide a callback_func, described below.) If the
+    primary backend name returned by the selector function doesn't correspond
+    to any registered backend, the function will raise a ``InvalidBackend``
+    exception. If any referenced secondary backends are not registered names,
+    they will be discarded and logged.
+
+    The selector function and callback function are both provided as paths to a
+    callable that will be imported at backend instantation.
+
+    Implementation notes:
+
+    - Only method access is delegated to the individual backends. Attribute
+      values are returned from the base backend. Only methods that are defined
+      on the base backend are eligble for delegation (since these methods are
+      considered the public API.)
+    - The backend makes no attempt to synchronize common backend option values
+      between backends (e.g. TSDB rollup configuration) to ensure equivalency
+      of request parameters based on configuration.
+    - Each backend is associated with an executor pool which defaults to a
+      thread pool implementation unless otherwise specified in the backend
+      configuration. If the backend itself is not thread safe (due to socket
+      access, etc.), it's recommended to specify a pool size of 1 to ensure
+      exclusive access to resources. Each executor is started when the first
+      task is submitted.
+    - The request is added to the request queue of the primary backend using a
+      blocking put. The request is added to the request queue(s) of the
+      secondary backend(s) as a non-blocking put (if these queues are full, the
+      request is rejected and the future will raise ``Queue.Full`` when
+      attempting to retrieve the result.)
+    - The ``callback_func`` is called after all futures have completed, either
+      successfully or unsuccessfully. The function is parameters are the method
+      name, calling arguments (as returned by ``inspect.getcallargs``) and a
+      mapping of backend name to ``Future`` objects.
+    """
+
+    class InvalidBackend(Exception):
+        """\
+        Exception raised when an invalid backend is returned by a selector
+        function.
+        """
+
+    def __init__(self, backend_base, backends, selector_func, callback_func=None):
```

Comment:
I'd like to find a way to test this, but the `import_string` bits (necessary for the way we invoke things using `LazyServiceWrapper` makes this tricky without adding dependencies on real services (e.g. `TSDB`.) The best approach I could think of was to mock out `import_string` with an identity function that so callables can be passed in instead. I'm not sure if anybody has any better ideas…?